src/context/controls-context.js

```diff
@@ -0,0 +1,110 @@
+import { createContext, useContext, useReducer } from "react";
+import {
+  colorMaps,
+  range,
+  initialTransferFunction,
+} from "../assets/config.json";
+
+const ControlsContext = createContext();
+
+// Custom component to provide the Controls context
+function ControlsProvider(props) {
+  const [state, dispatch] = useReducer(volumeReducer, {
+    // Constants
+    // TODO: MAKE_CAPITAL_CASE
+    allColorMaps: colorMaps,
+    sliderRange: range,
+    USE_TRANSFER_FUNCTION: true,
+
+    // Color Map and Opacity
+    colorMap: colorMaps[0],
+    transferFunctionNodes: initialTransferFunction,
+
+    // Sliders
+    xLowerBound: range.min,
+    xUpperBound: range.max,
+    yLowerBound: range.min,
+    yUpperBound: range.max,
+    zLowerBound: range.min,
+    zUpperBound: range.max,
+  });
+
+  const value = { state, dispatch };
+  return (
+    <ControlsContext.Provider value={value}>
+      {props.children}
+    </ControlsContext.Provider>
+  );
+}
+
+// Custom hook to get the current ControlsContext
+function useControlsContext() {
+  const context = useContext(ControlsContext);
+  if (context === undefined) {
+    throw new Error("useControls must be used within a ControlsProvider");
+  }
+  return context;
+}
+
+// Custom reducer to update the ControlsContext
+function volumeReducer(state, action) {
+  switch (action.type) {
+    case "CHANGE_COLOR_MAP": {
+      return {
+        ...state,
+        colorMap: action.payload,
+      };
+    }
+    case "CHANGE_TRANSFER_FUNCTION": {
+      return {
+        ...state,
+        transferFunctionNodes: action.payload,
+      };
+    }
+    case "CHANGE_X_SLIDER": {
+      return {
+        ...state,
+        xLowerBound: action.payload1,
```

Comment:
maybe make the payload for these cases an object like `{xUpperBound: 0.1, yUpperBound: 0.2}` then you can just splat the payload `...payload`

Reply:
then you could also collapse the cases so you only have one return

```js
case "CHANGE_X_SLIDER":
case "CHANGE_Y_SLIDER":
case "CHANGE_Z_SLIDER": {
  return { ...state, ...payload };
} 
```

Reply:
you would need to pass the correct keys into the payload `{xUpperBound: 0.1, yUpperBound: 0.2}` then you can just splat it in. So an array valued thing wouldn't work for that.

Reply:
Took a little JS trickery build and pass the variable name into the object but it's working now!